File: src/main/java/com/mikemunhall/katas/java/Fibonacci.java
package com.mikemunhall.katas.java;

import java.util.ArrayList;
import java.util.List;

public class Fibonacci {

    // 1, 2, 3, 5, 8, 13, 21, 34 ...
    public static List<Integer> getResult(int max) {
        List<Integer> acc = new ArrayList<>();
        Integer current = 0;

        while (current < max) {
            // [] or [1]
            if (acc.size() <= 1) {
                acc.add(1);
                current++;
                continue;
            }

            // [1, 1, ...]
            acc.add(acc.get(acc.size() - 1) + acc.get(acc.size() - 2));
            current++;
            continue;
        }

        return acc;
    }
}

File: src/main/scala/com/mikemunhall/katas/scala/Fibonacci.scala
package com.mikemunhall.katas.scala

object Fibonacci {

  // List(1, 1, 2, 3, 5, 8, 13, 21, 35, ...)
  def getResult(max: Int, acc: List[Int] = Nil): List[Int] = {

    if (max < 0) {
      throw new IllegalArgumentException
    } else if (max == 0) {
      acc
    } else if (max == 1) {
      List(1)
    } else if (acc == Nil) {
      getResult(max, List(1, 1))
    } else if (acc.size >= max) {
      acc.reverse
    } else {
      getResult(max, acc.head + acc.tail.head :: acc)
    }

  }
}
